### PR TITLE
fix multi-hop traverse with name path

### PR DESCRIPTION
**Pull request by nafraf**
This PR is to solve #2739 "Exact path length queries don't seem to respect the path length"

The problem seems to be that ```QGEdge_VariableLength()``` can't detect correctly that the edge has variable length when ```minHops == maxHops```, because it returns ```(e->minHops != e->maxHops)```.

In queries like the following,  there are variable length edges where ```minHops == maxHops```, and ```QGEdge_VariableLength()`` erroneously return false:
```
GRAPH.QUERY g "MATCH p = ({v:'a'})-[*2]-({v:'c'}) RETURN length(p)"
```
```
GRAPH.QUERY g "MATCH p = ({v:'a'})-[*2..2]-({v:'c'}) RETURN length(p)"
```

Then, the idea is to create a new property ```varLen``` that is ```true``` when the edge has variable length specified, without consider the values of minHops, maxHops or even if they are missing.

Comments are welcome.

### CODE REVIEW

File: tests/flow/test_variable_length_traversals.py [hunk 1]
```diff
@@ -182,3 +182,19 @@ def test10_filtered_edges_after_segment_change(self):
         actual_result = redis_graph.query(query)
         expected_result = [['A', 'B']]
         self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # Test variable-length edges using minHops and maxHops
+    def test11_variable_length_edges(self):
+        query = """CREATE (a {v:'a'}), (b {v:'b'}), (c {v:'c'}), (d {v:'d'}), (a)-[:R]->(b), (b)-[:R]->(c), (c)-[:R]->(a), (d)-[:R]->(d)"""
+        actual_result = redis_graph.query(query)
+        query_to_expected_result = {
+            "MATCH p = ({v:'a'})-[*2]-({v:'c'}) RETURN length(p)" : [[2]],
```
Comment (AviAvni) [inline tests/flow/test_variable_length_traversals.py:191]: this one is not variable length it fixed length

File: tests/flow/test_variable_length_traversals.py [hunk 2]
```diff
@@ -182,3 +182,19 @@ def test10_filtered_edges_after_segment_change(self):
...
+            "MATCH p = ({v:'a'})-[*2..]-({v:'c'}) RETURN length(p)" : [[2]],
+            "MATCH p = ({v:'a'})-[*2..2]-({v:'c'}) RETURN length(p)" : [[2]],
```
Comment (AviAvni) [inline tests/flow/test_variable_length_traversals.py:193]: this one is not variable length it fixed length

File: tests/flow/test_variable_length_traversals.py [hunk 3]
```diff
@@ -182,3 +182,19 @@ def test10_filtered_edges_after_segment_change(self):
...
+            "MATCH p = ({v:'a'})-[*]-({v:'c'}) RETURN length(p)" : [[2],[1]],
+            "MATCH p = ({v:'a'})-[*..]-({v:'c'}) RETURN length(p)" : [[2],[1]],
+            "MATCH p = ({v:'d'})-[*0]-() RETURN length(p)" : [[0]],
```
Comment (AviAvni) [inline tests/flow/test_variable_length_traversals.py:196]: this one is not variable length it fixed length

**Review by AviAvni — CHANGES_REQUESTED** [review]
see comments

File: tests/flow/test_variable_length_traversals.py (hunk 1 above)
Comment (nafraf) [reply tests/flow/test_variable_length_traversals.py:191]: You are right, it's fixed length.
Should I rename the ```QGEdge_VariableLength()``` method and the property ```varLen```? Because with the change of this PR, it detects if a range length was specified, fixed or variable.

And maybe the function ```NewCondVarLenTraverseOp()``` should be renamed too.

File: src/graph/query_graph.c [hunk 4]
```diff
@@ -118,11 +118,14 @@ static void _QueryGraphAddEdge
...
 	const cypher_astnode_t *range = cypher_ast_rel_pattern_get_varlength(ast_entity);
 	if(range) {
+		edge->varLen = true;
 		const cypher_astnode_t *start = cypher_ast_range_get_start(range);
 		const cypher_astnode_t *end = cypher_ast_range_get_end(range);
 		if(start) edge->minHops = AST_ParseIntegerNode(start);
 		if(end) edge->maxHops = AST_ParseIntegerNode(end);
 		else edge->maxHops = EDGE_LENGTH_INF;
+	} else {
+		edge->varLen = false;
```
[the remark this answers is not in the document]
Comment (nafraf) [reply src/graph/query_graph.c]: Thanks. Fixed in [ec8e720](https://github.com/RedisGraph/RedisGraph/pull/2743/commits/ec8e7207898811d247070ae9abe61a299789a22b)

File: tests/flow/test_variable_length_traversals.py (hunk 1 above)
Comment (nafraf) [reply tests/flow/test_variable_length_traversals.py:191]: I renamed some functions. Please check if you agree.
```QGEdge_RangeLength()``` to ```QGEdge_RangeLength()```
```_AlgebraicExpression_IsVarLen()``` to ```_AlgebraicExpression_IsRangeLen()```

Property ```varLen``` renamed to ```rangeLen```

The test were not removed, just renamed, for all tests  ```rangeLen == true```
```test11_variable_length_edges()``` to ```test11_range_length_edges()```

**Comment by swilly22** [comment]
There's a big performance difference between the `Conditional Traverse` operation and the `Conditional Variable Length Traverse`
The former performs traversal via matrix matrix multiplication (BFS) while the latter performs DFS while consulting with the matrix which is costlier.

In cases where the traversal length is constant (a:A)-[:R*n]->(z:Z) n hops, we want to produce the following algebraic expression:
`A * R^n * Z`

Please note that if the pattern is named: e.g. `P = (a:A)-[:R*n]->(z:Z)` then we fall back to DFS as we need each component of the path.

**Comment by swilly22** [comment]
Following up on this, the original issue:
```
GRAPH.QUERY g "CREATE (a{v:'a'}),(b{v:'b'}),(c{v:'c'}),(a)-[:R]->(b),(b)-[:R]->(c),(c)-[:R]->(a)"
GRAPH.QUERY g "MATCH p = ({v:'a'})-[*2]-({v:'c'}) RETURN length(p)"
```
Creates a graph where `c` can be reached from `a` in multiple ways:
1.` a->b->c`
2. `a<-c`

The request in the READ query is to get from `a` to `c` via 2 hops `({v:'a'})-[*2]-({v:'c'})`
The only path which will get us there is: ` a->b->c`

The appropriate Algebraic expression which denotes this pattern is: `THE_ADJ` * `THE_ADJ` applied by a single
`Conditional Traverse` operation.

```
GRAPH.explain g "MATCH p = ({v:'a'})-[*2]->({v:'c'}) RETURN p"
1) "Results"
2) "    Project"
3) "        Filter"
4) "            Conditional Traverse | (@anon_0)-[@anon_2*2..2]->(@anon_1)"
5) "                Filter"
6) "                    All Node Scan | (@anon_0)"
```

The problem here is that `Conditional Traverse` performs traversal via matrix matrix multiplication which "losses" traversed edges along the path. after the operation finishes we're left with the following information:
1. src node
2. dest node
3. last traversed edge

This lost of path critical information makes the Path object construction not possible.

There are two solutions as we see it:
1. when dealing with named paths of length > 1 always use DFS (`Conditional Variable Length Traverse`)
2. use multiple `Conditional Traverse` operations one for each hop along the constant length path.

This PR seems to have taken the first approach which is OK, but performance wise we might want to use multiple `Conditional Traverse` operations if the number of hops is say less than 5.

File: tests/flow/test_variable_length_traversals.py [hunk 5]
```diff
@@ -182,3 +182,19 @@ def test10_filtered_edges_after_segment_change(self):
         actual_result = redis_graph.query(query)
         expected_result = [['A', 'B']]
         self.env.assertEquals(actual_result.result_set, expected_result)
+
+    # Test range-length edges
+    def test11_range_length_edges(self):
```
Comment (swilly22) [inline tests/flow/test_variable_length_traversals.py:187]: Quick re-write
```python
     # Test range-length edges
     def test11_range_length_edges(self):
         # clear previous data
         conn = self.env.getConnection()
         conn.flushall()

         # populate graph
         # create a graph with 4 nodes
         # a->b
         # b->c
         # c->a
         # d->d
         query = """CREATE (a {v:'a'}), (b {v:'b'}), (c {v:'c'}), (d {v:'d'}),
                           (a)-[:R]->(b), (b)-[:R]->(c), (c)-[:R]->(a), (d)-[:R]->(d)"""

         actual_result = redis_graph.query(query)

         # validation queries
         query_to_expected_result = {
             "MATCH p = (a {v:'a'})-[*2]-(c {v:'c'}) RETURN length(p)" : [[2]],
             "MATCH p = (a {v:'a'})-[*2..]-(c {v:'c'}) RETURN length(p)" : [[2]],
             "MATCH p = (a {v:'a'})-[*2..2]-(c {v:'c'}) RETURN length(p)" : [[2]],
             "MATCH p = (a {v:'a'})-[*]-(c {v:'c'}) WITH length(p) AS len RETURN len ORDER BY len" : [[1],[2]],
             "MATCH p = (a {v:'a'})-[*..]-(c {v:'c'}) WITH length(p) as len RETURN len ORDER BY len" : [[1],[2]],
             "MATCH p = (d {v:'d'})-[*0]-() RETURN length(p)" : [[0]],
         }

         # validate query results
         for query, expected_result in query_to_expected_result.items():
             actual_result = redis_graph.query(query)
             self.env.assertEquals(actual_result.result_set, expected_result)
```

File: src/graph/entities/qg_edge.h [hunk 6]
```diff
@@ -21,6 +21,7 @@ struct QGEdge {
 	uint maxHops;           // maximum number of hops this edge represents
 	bool bidirectional;     // edge doesn't have a direction
 	bool shortest_path;     // only edges in the shortest paths should be collected
+	bool rangeLen;          // this edge represents a relationship with range of hops: minHops, maxHops or both of them were defined
```
Comment (swilly22) [inline src/graph/entities/qg_edge.h:24]: Seems to be redundant please deduct from `minHops` and `maxHops`

File: src/graph/entities/qg_edge.h [hunk 7]
```diff
@@ -56,8 +57,8 @@ QGEdge *QGEdge_Clone
 	const QGEdge *orig
 );
 
-// determine whether this is a variable length edge
-bool QGEdge_VariableLength
+// determine whether this is a edge with a range of length defined (variable length [*minHops..maxHops] or fixed length [*Hops])
+bool QGEdge_RangeLength
```
Comment (swilly22) [inline src/graph/entities/qg_edge.h:61]: Please replace with the following 2 functions:
1. `QGEdge__VariableLength` which simply returns true if `minHops ` != `maxHops`
2.`QGEdge_RangeLength` which simply returns true if `minHops` > 1

Please make sure patterns such as `()-[*0]-()` are still supported and are functioning as expected.

File: src/graph/query_graph.c [hunk 8]
```diff
@@ -118,6 +118,7 @@ static void _QueryGraphAddEdge
 	// incase of a variable length edge, set edge min/max hops
 	const cypher_astnode_t *range = cypher_ast_rel_pattern_get_varlength(ast_entity);
 	if(range) {
+		edge->rangeLen = true;
```
Comment (swilly22) [inline src/graph/query_graph.c:121]: Redundant please see comments in `qg_edge.h`

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c [hunk 9]
```diff
@@ -44,7 +44,7 @@ static inline bool _should_populate_edge
 	QGEdge *e
 ) {
 	ASSERT(e != NULL);
-	return (_referred_entity(e->alias) || QGEdge_VariableLength(e));
```
Comment (swilly22) [inline src/arithmetic/algebraic_expression/algebraic_expression_construction.c:47]: an edge taking part in a `named-path` e.g. `MATCH P = (a)-[:R]->(b)` should be marked as a referenced entity.
As such with the given comment on `qg_edge.h` I believe we can revert this change back to its original version.

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c [hunk 10]
```diff
@@ -70,7 +70,7 @@ static bool _AlgebraicExpression_ContainsVariableLengthEdge
 	case AL_OPERAND:
 		if(exp->operand.edge) {
 			QGEdge *e = QueryGraph_GetEdgeByAlias(qg, exp->operand.edge);
-			return QGEdge_VariableLength(e);
+			return QGEdge_RangeLength(e);
```
Comment (swilly22) [inline src/arithmetic/algebraic_expression/algebraic_expression_construction.c:73]: Given the comment on `qg_edge.h` I believe we can revert this change back to its original version.

File: src/arithmetic/algebraic_expression/algebraic_expression_construction.c [hunk 11]
```diff
@@ -334,7 +334,7 @@ static AlgebraicExpression *_AlgebraicExpression_OperandFromEdge
 	AlgebraicExpression  *add               =  NULL;
 	AlgebraicExpression  *root              =  NULL;
 	AlgebraicExpression  *src_filter        =  NULL;
-	bool                 var_len_traversal  =  QGEdge_VariableLength(e);
+	bool                 var_len_traversal  =  QGEdge_RangeLength(e);
```
Comment (swilly22) [inline src/arithmetic/algebraic_expression/algebraic_expression_construction.c:337]: Given the comment on qg_edge.h I believe we can revert this change back to its original version.

File: src/execution_plan/execution_plan_build/build_match_op_tree.c [hunk 12]
```diff
@@ -108,7 +108,7 @@ static void _ExecutionPlan_ProcessQueryGraph
 					QueryGraph_GetEdgeByAlias(qg, AlgebraicExpression_Edge(exp));
 			}
 
-			if(edge && QGEdge_VariableLength(edge)) {
+			if(edge && QGEdge_RangeLength(edge)) {
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/build_match_op_tree.c:111]: due to the comment on `qg_edge.h` please update this condition to:
```c
if(edge && (QGEdge_VariableLength(edge) || QGEdge_RangeLength(edge))) { ...
```

File: src/execution_plan/optimizations/traverse_order_utils.c [hunk 13]
```diff
@@ -272,8 +272,8 @@ void TraverseOrder_ScoreExpressions
 			score = TraverseOrder_FilterExistenceScore(exp, qg,
 													   filtered_entities);
 			if(score > 0) {
-				if(_AlgebraicExpression_IsVarLen(exp, qg)) {
-					// variable length traversal should always "lose" to its
+				if(_AlgebraicExpression_IsRangeLen(exp, qg)) {
```
Comment (swilly22) [inline src/execution_plan/optimizations/traverse_order_utils.c:275]: I believe this entire file should be reverted once the change to `qg_edge.h` is applied.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/graph/entities/qg_edge.h [hunk 14]
```diff
@@ -62,6 +62,12 @@ bool QGEdge_VariableLength
 	const QGEdge *e
 );
 
+// determine whether this is a edge with a range of length defined (variable length [*minHops..maxHops] or fixed length [*Hops])
```
Comment (AviAvni) [inline src/graph/entities/qg_edge.h:65]: I think comment need update here

Comment (nafraf) [reply src/graph/entities/qg_edge.h]: When I was updating the comment I noticed that maybe this function should be replaced by a ```QGEdge_LengthOne()```. See comment below.

**Comment by nafraf** [comment]
Thanks for the comments and description of the algorithm. Now the code is more clear to me.

With the first approach, to support the case ```()-[*0]-()``` I had to use the ```Conditional Variable Length Traverse```, then I think that ```QGEdge_RangeLength()``` should be replaced by a function ```QGEdge_LengthOne()``` which returns true if length is one. What do you think?

**Review by AviAvni — COMMENTED** [review]
rename function
